## Driving right (1)

¿Recuerda el dataset *cars* que contiene los automóviles por cada 1000 personas (cars_per_cap) y si las personas conducen correctamente (drives_right) para diferentes países (country)? El código que importa estos datos en formato CSV en Python como DataFrame está disponible.

Comencemos de manera simple e intentemos encontrar todas las observaciones en *cars* donde *drives_right* son *True*.

*drives_right* es una columna booleana, por lo que tendrá que extraerla como una Serie y luego usar esta Serie booleana para seleccionar observaciones *cars*.

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Extraiga la columna *drives_right* como una serie de pandas y guárdela como *dr*.
* Utilice *dr*, una serie booleana, para subconjugar el DataFrame *cars*. Almacene la selección resultante en *sel*.
* Imprima *sel* y afirme que *drives_right* son *True* para todas las observaciones.

In [2]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Extract drives_right column as Series: dr
dr = cars["drives_right"]

# Use dr to subset cars: sel
sel = cars[dr]

# Print sel
print(sel)

     cars_per_cap        country  drives_right
US            809  United States          True
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


## Driving right (2)

El código en el ejemplo anterior funcionó bien, pero en realidad creó innecesariamente una nueva variable *dr*. Puede lograr el mismo resultado sin esta variable intermedia. Ponga el código que computa *dr* directamente entre corchetes que seleccionan las observaciones *cars*.

<div class="alert alert-info">
    
**Actividad**
    
</div>

Convierta el código de la derecha a una línea que calcule la variable *sel* como antes.

In [3]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Convert code to a one-liner
sel = cars[cars['drives_right']]

# Print sel
print(sel)

     cars_per_cap        country  drives_right
US            809  United States          True
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


## Cars per capita (1)

Sigamos con los datos *cars* un poco más. Esta vez, desea saber qué países tienen una cifra alta de automóviles per cápita . En otras palabras, en qué países muchas personas tienen un automóvil, o tal vez varios automóviles.

De manera similar al ejemplo anterior, querrá construir una Serie booleana, que luego puede usar para subconjugar el DataFrame *cars* para seleccionar ciertas observaciones. Si quieres hacer esto de una sola vez, ¡está perfectamente bien!

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Seleccione la columna *cars_per_cap* de *cars* como Serie de pandas y guárdela como *cpc*.
* Use *cpc* en combinación con un operador de comparación y 500. ¿Quieres terminar con una serie booleana que *True* es si el país correspondiente tiene más de 500 *cars_per_cap* y *False* de lo contrario. Almacene esta serie booleana como *many_cars*.
* Use *many_cars* para subconjunto *cars*, similar a lo que hizo antes. Almacene el resultado como *car_maniac*.
* Imprima *car_maniac* para ver si lo hizo bien.

In [6]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Create car_maniac: observations that have a cars_per_cap over 500
car_maniac = cars[cars["cars_per_cap"]>500]

# Print car_maniac
print(car_maniac)

     cars_per_cap        country  drives_right
US            809  United States          True
AUS           731      Australia         False
JPN           588          Japan         False


## Cars per capita (2)

Recuerdan los *np.logical_and(), np.logical_or() y np.logical_not()*, las operadores variantes de Numpy para and, or y not? También puede usarlos en Pandas Series para realizar operaciones de filtrado más avanzadas.

Tome este ejemplo que selecciona las observaciones que tienen *cars_per_cap* entre 10 y 80. Pruebe estas líneas de código paso a paso para ver qué está sucediendo.

```python
cpc = cars['cars_per_cap']  
between = np.logical_and(cpc > 10, cpc < 80)  
medium = cars[between]
```

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Utilice el ejemplo de código anterior para crear un DataFrame *medium*, que incluye todas las observaciones de *cars* que tienen un intervalo *cars_per_cap* entre *100* y *500*.
* Imprimir *medium*.

In [8]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Import numpy, you'll need this
import numpy as np

# Create medium: observations with cars_per_cap between 100 and 500
medium = cars[np.logical_and(cars['cars_per_cap']>100,cars['cars_per_cap']<500)]

# Print medium
print(medium)

    cars_per_cap country  drives_right
RU           200  Russia          True


## Bucle sobre DataFrame (1)

Iterar sobre un Pandas DataFrame generalmente se realiza con el método **iterrows()**. Utilizado en un bucle *for*, cada observación se repite una y otra vez, la etiqueta de la fila y el contenido real de la fila están disponibles:

```python
for lab, row in brics.iterrows() :
    ...
```    
En este y en los siguientes ejercicios trabajará con el DataFrame *cars*. Contiene información sobre los automóviles per cápita y si las personas conducen hacia la derecha o hacia la izquierda en siete países del mundo.

<div class="alert alert-info">
    
**Actividad**
    
</div>

+ Escriba un bucle *for* que itera sobre las filas de *cars* y en cada iteración realice dos llamadas **print()**: una para imprimir la etiqueta de la fila y otra para imprimir todos los contenidos de las filas.

In [ ]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Iterate over rows of cars
for lab, row in cars.iterrows():
    print(lab)
    print(row)

## Bucle sobre DataFrame (2)

Los datos de fila generados por cada ejecución de **iterrows()** son una serie de Pandas. Este formato no es muy conveniente para imprimir. Afortunadamente, puede seleccionar fácilmente variables de la Serie Pandas usando corchetes:

```python
for lab, row in brics.iterrows():  
    print(row['country'])
```

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Usando los iteradores *lab* y *row*, adapte el código en el bucle for de modo que se imprima la primera iteración "*US: 809*", la segunda iteración "*AUS: 731*", etc.
* La salida debe estar en el formulario "*country: cars_per_cap*". Asegúrese de imprimir esta cadena exacta (con el espacio correcto).
    * Puede usar *str()* para convertir sus datos enteros en una cadena para poder imprimirlos junto con la etiqueta del país.


In [ ]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Adapt for loop
for lab, row in cars.iterrows() :
    print(lab + ": " + str(row["cars_per_cap"]))
    # print(row)

## Agregar columna (1)

En el siguiente ejemplo, se agregará la longitud de los nombres del país en el DataFrame *brics* en una nueva columna:

```python
for lab, row in brics.iterrows() :  
    brics.loc[lab, "name_length"] = len(row["country"])
```    
Puede hacer cosas similares en el DataFrame *cars*.

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Use un bucle *for* para agregar una nueva columna, llamada *COUNTRY*, que contenga una versión en mayúscula de los nombres de países en la columna "*country*". Puede usar el método de string **upper()** para esto.
* Para ver si su código funcionó, imprima *cars*. No indente este código para que no forme parte del bucle *for*.

In [ ]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Code for loop that adds COUNTRY column
for lab, row in cars.iterrows():
    cars.loc[lab,"COUNTRY"] = str.upper(row["country"])

# Print cars
print(cars)

Agregar columna (2)
--

Usar **iterrows()** para iterar sobre cada observación de un Pandas DataFrame es fácil de entender, pero no muy eficiente. En cada iteración, estás creando una nueva serie Pandas.

Si desea agregar una columna a un DataFrame llamando a una función en otra columna, el método **iterrows()** en combinación con un bucle *for* no es la mejor opción. En cambio, querrás usar **apply()**.

Compare la versión **iterrows()** con la versión **apply()** para obtener el mismo resultado en el DataFrame *brics*:

```python
for lab, row in brics.iterrows() :  
    brics.loc[lab, "name_length"] = len(row["country"])

brics["name_length"] = brics["country"].apply(len)
```
Podemos hacer algo similar para llamar al método **upper()** en cada nombre de la columna *country*. Sin embargo, **upper()** es un método, por lo que necesitaremos un enfoque ligeramente diferente:

<div class="alert alert-info">
    
**Actividad**
    
</div>

* Reemplace el bucle *for* con una línea que usa **.apply(str.upper)**. La llamada debe dar el mismo resultado: se debe agregar una columna *COUNTRY* en mayúscula de los nombres de los países contenida en *cars*.
* Como de costumbre, imprima para *cars ver los frutos de su arduo trabajo.

In [ ]:
# Import cars data
import pandas as pd
cars = pd.read_csv('data/cars.csv', sep =';', index_col=0)

# Use .apply(str.upper)
cars["COUNTRY"] = cars["country"].apply(str.upper)
print(cars)